compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -29,31 +29,31 @@ impl DeduceReadOnly {
 }
 
 impl<'tcx> Visitor<'tcx> for DeduceReadOnly {
-    fn visit_local(&mut self, local: Local, mut context: PlaceContext, _: Location) {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _location: Location) {
         // We're only interested in arguments.
-        if local == RETURN_PLACE || local.index() > self.mutable_args.domain_size() {
+        if place.local == RETURN_PLACE || place.local.index() > self.mutable_args.domain_size() {
             return;
         }
 
-        // Replace place contexts that are moves with copies. This is safe in all cases except
-        // function argument position, which we already handled in `visit_terminator()` by using the
-        // ArgumentChecker. See the comment in that method for more details.
-        //
-        // In the future, we might want to move this out into a separate pass, but for now let's
-        // just do it on the fly because that's faster.
-        if matches!(context, PlaceContext::NonMutatingUse(NonMutatingUseContext::Move)) {
-            context = PlaceContext::NonMutatingUse(NonMutatingUseContext::Copy);
-        }
-
-        match context {
-            PlaceContext::MutatingUse(..)
-            | PlaceContext::NonMutatingUse(NonMutatingUseContext::Move) => {
+        let mark_as_mutable = match context {
+            PlaceContext::MutatingUse(..) => {
                 // This is a mutation, so mark it as such.
-                self.mutable_args.insert(local.index() - 1);
+                true
+            }
+            PlaceContext::NonMutatingUse(NonMutatingUseContext::AddressOf) => {
+                // Whether mutating though a `&raw const` is allowed is still undecided, so we
+                // disable any sketchy `readonly` optimizations for now.
+                // But we only need to do this if the pointer would point into the argument.
+                !place.is_indirect()
```

Comment:
Why is it okay to consider indirect places immutable? I have no idea what kind of code this affects, but it seems very suspicious to me that whether or not the place is indirect would make any difference.

Reply:
If you have

```rust
fn foo(arg: *const u8) { /* ... */ }
```

then `addr_of!(*arg)` can't change the value of `arg`, only `addr_of!(arg)` can.

In practice, this doesn't actually make a difference and we could always return true here, because currently
1. single pointers are never passed indirectly
2. when accessing a pointer in a compound type, e.g. `addr_of!(*(arg.ptr))` this operation is split in two mir statements `let tmp = deref_copy arg.ptr` and `&raw const *tmp` and so this `&raw const` already doesn't affect whether `arg` is readonly.

Reply:
> then addr_of!(*arg) can't change the value of arg, only addr_of!(arg) can.

This should apply uniformly to `&[raw] [mut|const]` (including regular non-raw borrows) then, right?

Reply:
That seems to be correct. In that case I'd probably just remove the special case here (and just return true) rather than introduce more special casing, because it doesn't actually matter.

Reply:
Sounds good to me.